[`pylint`] Fix false negatives of `chained-comparison` (`PLR1716`)

Bnyro opened the pull request:
<!--
Thank you for contributing to Ruff/ty! To help us out with reviewing, please consider the following:

- Does this pull request include a summary of the change? (See below.)
- Does this pull request include a descriptive title? (Please prefix with `[ty]` for ty pull
  requests.)
- Does this pull request include references to any relevant issues?
-->

## Summary
All the following if cases have the same semantic meaning:
```py
x = 5

if 4 < x and x < 7:
    ...

if x > 4 and x < 7:
    ...

if x > 4 and 7 > x:
    ...
```
However, the `boolean-chained-comparison` rule currently only spots the first case, even though all of them could be simplified by comparison chaining.

This PR implements a detection mechanism for the missing cases, however it does not implement quick fixes for them since this turned out to be quite difficult.

<!-- What's the purpose of the change? What does it do, and why? -->

## Test Plan

<!-- How was it tested? -->
I've added new test case snapshots and compared the behavior to `pylint`.

Bnyro committed:
[`ruff`] Fix false negatives of `chained-comparison` (PLR1716)

ntBre reviewed (commented):
Thanks for working on this! The new fix looks pretty complicated, so I still need to take a closer look, but I noticed a couple of issues in passing.

I'll also quickly close and reopen the PR to re-trigger CI. Not sure what happened earlier, but it looks like it got stuck for some reason.

ntBre commented on crates/ruff_linter/src/rules/pylint/rules/boolean_chained_comparison.rs:
````diff
@@ -35,20 +35,28 @@ use crate::{AlwaysFixableViolation, Edit, Fix};
 /// ```
 #[derive(ViolationMetadata)]
-#[violation_metadata(stable_since = "0.9.0")]
+#[violation_metadata(preview_since = "0.15.1")]
````
Sorry for the delay on review. I still need to do a full review, but I noticed this in passing today.

This isn't what I meant by preview change. We don't need to move the rule out of the stable group back into preview, which is what this change would do. Instead we need to add a preview function like the ones in the file I linked to and preview-gate just the new behavior of the rule.

MichaReiser commented on crates/ruff_linter/src/rules/pylint/rules/boolean_chained_comparison.rs:
```diff
@@ -70,88 +111,158 @@ pub(crate) fn boolean_chained_comparison(checker: &Checker, expr_bool_op: &ExprB
-        let Expr::Name(right_compare_left) = &*right_compare.left else {
-            continue;
-        };
+        // first case: all comparison operations are pointing in the same direction
+        // e.g. <expr> <=? x and x <=? <expr>
+        if left_chain_direction == right_chain_direction {
+            if is_chained_comparison_reorder_enabled(checker.settings())
+                && let Expr::Name(left_compare_left) = &*left_compare.left
+                && let Some(Expr::Name(right_compare_right)) = right_compare.comparators.last()
+                && left_compare_left.id() == right_compare_right.id()
+            {
+                // could be simplified by swapping the condition parts
+                // e.g. x < 5 and 3 < x -> 3 < x < 5
+                // auto-fix would be possible but is not implemented yet
+                checker.report_diagnostic(
+                    BooleanChainedComparison {
+                        lhs_var_position: ComparisonReusedVariablePosition::Left,
+                        rhs_var_position: ComparisonReusedVariablePosition::Right,
+                    },
+                    TextRange::new(left_compare.start(), right_compare.end()),
+                );
+                continue;
+            }
```
You can add a test using `assert_diagnostics_diff` where it compares preview and non preview behavior, see https://github.com/astral-sh/ruff/blob/main/crates/ruff_linter/src/rules/fastapi/mod.rs#L35-L54 (you don't need to use `test_case` if you only have a single test file)

Bnyro replied:
Neat that we can simply compare the diffs like that, thank you for the hint!

MichaReiser commented:
I found two more semantic issues

```py
a = 1
b = 2

if a < b and a > b:
    pass
```

This now creates two diagnostics instead of just one.

```py
if a < b and c >= b:
    pass
```

We don't show the correct suggestion in this case:

```
warning: Detected debug build without --no-cache.
error[PLR1716]: Contains chained boolean comparison that can be simplified
  --> /private/tmp/plr1716_finding3.py:9:4
   |
 7 | c = 3
 8 |
 9 | if a < b and c >= b:
   |    ^^^^^^^^^^^^^^^^
10 |     pass
   |
help: Use a single compare expression. For example, '3 > x and 7 >= x' could be simplified to '3 < x <= 7'.

Found 1 error.
```

Bnyro commented:
Thank you for testing this @MichaReiser, I've added both your examples as test cases and fixed them.

Additionally, we're now displaying the correct comparison direction in the title message, e.g. we're always displaying `x < 7` for everything that's like `var < expr` instead of possibly also displaying `x > 7` and vice-versa.
